torch/utils/data/sampler.py

```diff
@@ -333,26 +334,18 @@ def __init__(
 
     def __iter__(self) -> Iterator[List[int]]:
         # Implemented based on the benchmarking in https://github.com/pytorch/pytorch/pull/76951
+        sampler_iter = iter(self.sampler)
         if self.drop_last:
-            sampler_iter = iter(self.sampler)
+            # Create multiple references to the same iterator
+            args = [sampler_iter] * self.batch_size
+            for batch in zip(*args):
+                yield list(batch)
+        else:
             while True:
-                try:
-                    batch = [next(sampler_iter) for _ in range(self.batch_size)]
-                    yield batch
-                except StopIteration:
+                batch: List[int] = list(itertools.islice(sampler_iter, self.batch_size))  # type: ignore[no-redef]
```

Comment:
We can just call it something else to remoe the redef typing error? No reason to reuse the name if we don't need to.

Reply:
That makes a lot of sense. Thanks!